### ISSUE TITLE
SPA mode how to's

### ISSUE THREAD

**Issue by PrimozRome** [nuxt/nuxt.js#1768]
I am opening this issue to resolve all the mysteries around the `SPA` mode. Seems like I am not the only one having problems with `spa` mode.

First of all I do not require `ssr` that's why I switched it off. 

# SPA mode setup
First of all I would like to add instructions how to setup `spa` mode. There are no instructions in documentation how this is done correctly. I am still not sure if I am doing it right but based on some other issues opened here I came to this:

1. Add `mode: 'spa'` to the `nuxt.config.js`
2. Modify `package.json` and change `scripts` key to this:
```
"scripts": {
    "dev": "nuxt --spa",
    "build": "nuxt build --spa",
    "start": "nuxt start",
    "generate": "nuxt generate",
    "lint": "eslint --ext .js,.vue --ignore-path .gitignore .",
    "precommit": "npm run lint"
  },
```
3. run `npm run dev` and navigate to `localhost:3000`

Can someone confirm this is the right setup?

# SPA mode issues and questions

**ISSUE  1** 404 page not found. For some reasons some of the pages in the get 404 error. I have opened issue > https://github.com/nuxt/nuxt.js/issues/1767. If I switch to universal mode page is working. There is another possible issue related to this > https://github.com/nuxt/nuxt.js/issues/1661

**QUESTION  1** Are the page and layout middleware(s) executed in the `spa` mode on the client side?  Are they executed on page refresh?

**QUESTION  2** - Are `asyncData()` or `fetch()` the right places to call backend API to get page data or should we use `mounted()` or `created()` hook to call the API?

**QUESTION 3** - Where to add authentication checks in `spa` mode? Should this be added as a registered NuxtJS plugin and registered in `nuxt.config.js` or should this be part of the page/layout middleware?

Thanks for answers.

<div align="right"><sub><em>This question is available on <a href="https://nuxtjs.cmty.io">Nuxt.js</a> community (<a href="https://nuxtjs.cmty.io/nuxt/nuxt.js/issues/c1588">#c1588</a>)</em></sub></div>

**Comment by jordanranson** [comment]
After changing my mode to spa, none of my pages will load anymore. The page loading bar shows up but none of the Axios requests in the pages asyncData methods are being fired.

**Comment by cretueusebiu** [comment]
If you add `mode: 'spa'` you don't need to include `--spa` in your scripts.

1. The page middleware are executed in spa mode on the client side.
2. Use `asyncData` and `fetch` if you want to fetch data before loading the component. If you want to load the page and then load the data you can make the call in the `created` hook. 
3. In your `nuxt.config.js` file you should have a global middleware that checks if a token exists (cookie/localstorage) and fetch the user from the server. Then in your pages you can have a middleware for guests and authenticated.

**Comment by PrimozRome** [comment]
@cretueusebiu thank you for clear explanations.

What about the [404 issue](https://github.com/nuxt/nuxt.js/issues/1767) on some routes in the spa mode?

**Comment by cretueusebiu** [comment]
Might just be something related to your project. Can you post a repro project ?

**Comment by PrimozRome** [comment]
@cretueusebiu can't unfortunately because it's a private repo. But my routes are set like this:

/layout/auth.vue - no layout middleware

```
<template>
  <div class="full-height">
    <nuxt/>
  </div>
</template>

<script>
export default {
}
</script>
```

/pages/auth.vue - no page middleware, just html in the template
/pages/auth/login - no page middleware, just html in the template with login form

Both routes `localhost:3000/auth` and `localhost:3000/auth/login` work in `ssr` mode but get `404` in `spa` with error described on the opened issue https://github.com/nuxt/nuxt.js/issues/1767

**Comment by PrimozRome** [comment]
@cretueusebiu you are right 404 is the issue of my application. I am trying to load a static asset in auth page, but in `spa` mode it doesn't work. Can you check the issue for details?

**Comment by dohomi** [comment]
I'd like to now what is the recommended debug for SPA mode. I am using `npm run dev` without issues, `npm run build` builds all files and creates the dist folder. But as soon I try to run or deploy the dist/index.html I only see the SPA loading spinner and nothing else happens. No error message, inside of network-panel of Chrome no activity, simply nothing. Is there any reference about the main differences? I am updating a universal project into SPA but currently stuck when trying to deploy that project. It would be great if `do / don't` and main differences would be documented. Thanks

**Comment by PrimozRome** [comment]
@cretueusebiu I have few more question in the SPA only mode. How does `env` work in `SPA`only mode? I am using [axios-module](https://github.com/nuxt-community/axios-module) and I tried this:

nuxt.config.js
```
  env: {
    BASE_URL: 'http://someurl/api/v1',
  },

  axios: {
    baseURL: process.env.BASE_URL,
  }

```

But when I start `npm run dev` I get this error:

```
TypeError: Cannot read property 'substr' of undefined
    at ModuleContainer.nuxtAxios (/Users/primozrome/Documents/Web/webshopx/node_modules/@nuxtjs/axios/lib/index.js:38:46)
```

If I change to this it works fine...

```
  axios: {
    baseURL: 'http://someurl/api/v1',
  }

```

**Comment by cretueusebiu** [comment]
I've posted a quick guide on setting up dotenv with Nuxt, just search in the closed issues.

**Comment by SteveALee** [comment]
I cannot deploy the /dist folder and get it to run.

I followed the nuxt installation notes, added `mode:'spa` to the config fule, installed and built and then copied the /dist folder to the root of a domain. All files appear to load in the F12 network tab but chrome and firefox show the page could not be found.

Have I missed something?

Also is it possible to specify a folder rather than the root of a domain?

**Comment by Elfayer** [comment]
*I can create a new issue if you prefer, I just thought it seems appropriate here.*

I have an initialization issue using Nuxt on SPA mode.

### Context
I have initialization in layout `default.vue` and page`index.vue`. In `default.vue`, I load from the server all the *global* information used by *all* pages, then each page should load its own specific information.

The thing is that `index.vue` page needs the data from `default.vue` loaded before loading itself.
On SPA mode, the ways to load I am aware of are the following:

1. middleware (executes on each routing AND first load (SPA mode specificity))
2. plugins (execute once before app loads)
3. fetch (to init `$store` with server data, has to be on a page, called before Vue creates the component)
4. asyncData (to init `data` with server data, has to be on a page, called before Vue creates the component)

That's it. Correct me if I am wrong.

### Question
**How do I load global app data FIRST and specific page data THEN?**

Now the problem is:
1. I don't want to load my app *global* information on each routing obviously, so out of my way.
2. Yes that should work! But I feel weird loading store *global* information from here... Isn't it supposed to be used for external libraries? **Caution:** Actually doesn't work, plugins are called after middlewares. I have an auth middleware, so no server call goes through in my plugins.
3. Yes, but no. It would work *if* I could fetch on `default.vue`(layout) and use `beforeCreate` on pages, but it's exactly the opposite of what is available. I can *only* call `fetch` on a page, not on a layout. And I need a *global* load for *each* page, whatever which page is requested first by the user (F5 use case). So I need something above the page load, to get application state and page state.
4. Nop, I want to load the store initial states.

### More (After some trials and thoughts)

I first thought that under universal mode (instead of SPA), you'd use a middleware to load your data on the server side. And it seems like under spa mode, plugins comes first to mind...

The problem with this idea: I want to load *once* and for all. A middleware is going to be executed on each routing, for SPA *or* universal mode. How am I supposed to load only once my store initial state? Specify the middleware on a page? Yes, but that doesn't work! I need something before any page loads and for any first load, whichever the page requested.

Maybe I am doing too many things in my layout and I should load everything outside of it. I don't really use layouts. I only have the two default ones: `default.vue` and `error.vue`. I'm not a fan of it, I feel it is an unnecessary layer that is only useful to avoid repeating the Navbar component on every page.

I am thinking of storing any store `action` promises in the store so that I can do something like: `this.$store.state.auth.onLoad.then()`, so I can wait for a Vuex module to be loaded before loading more. Or init the app on a global middleware and each store module should have a boolean `alreadyLoaded`...
But I am thinking Nuxt should have a solution to this.

Maybe I am misunderstanding something.
Any help/thoughts is appreciated ! :wink:

**Comment by jamesAtcodeninja** [comment]
Is it possible to use SPA and generate together to generate some routes?

**Comment by webdawe** [comment]
Hi Guys,
we would like to use nuxt ssr to serve search engine bots and need to use spa for the customers.
any suggestions?
need two apps and differentiate whether its search engine or suggest best way of doing it

Any thoughts appreciated!
Anil

**Comment by dons20** [comment]
@SteveALee Did you manage to find a solution to this issue?

**Comment by SteveALee** [comment]
@dons20 not that I recall :(
